# 🐍 4 — Iterators

**Welcome!** 🎉 This is a special topic that unlocks how loops *really* work in Python — and how programs handle **huge amounts of data** without running out of memory. It's a favourite interview question too!

Every topic is in **simple English**, with **real-life examples** and **line-by-line comments**.

---

## 📚 What you will learn here

| # | Topic |
|---|-------|
| 1 | What is iteration? |
| 2 | What is an iterator? (and why it saves memory) |
| 3 | Iterable vs Iterator vs Iteration |
| 4 | How to check if something is iterable / an iterator |
| 5 | How a `for` loop *really* works |
| 6 | Building your own `for` loop |
| 7 | Building your own `range()` |

---

## 1️⃣ What is Iteration?

**Iteration** is the process of going through a **sequence of data one item at a time**. Every time you use a loop, you are iterating — taking each item out one by one and doing something with it.

In [ ]:
num = [1, 2, 3]
# Going through the list one item at a time = iteration
for i in num:
    print(i)

1
2
3


## 2️⃣ What is an Iterator?

An **iterator** is the object that *makes* iteration happen. When you run a loop, an iterator secretly works behind the scenes, fetching items **one at a time** and handing them to the loop.

### 💡 The superpower: memory efficiency

An iterator does **not** load all the data into memory at once. It brings **one item** into memory, does the work, throws it away, then brings the next — so only **one item** is in memory at a time. This saves a *huge* amount of memory.

Let's prove it: show the numbers 1 to 100,000 in two ways and compare memory.

In [ ]:
import sys

# Way 1: build a LIST of all the numbers (loads everything into memory)
l = [x for x in range(1, 100000)]
sys.getsizeof(l) / 1024   # size in kilobytes

782.2109375

In [ ]:
# Way 2: use range() (an iterator — generates numbers on demand)
x = range(1, 100000)
sys.getsizeof(x) / 1024

0.046875

😲 **Look at the difference!** The list took ~**782 KB**, but `range` took only ~**0.05 KB**.

Why? The list stored all 100,000 numbers in memory at once. But `range` stored **nothing** — it created an iterator that produces each number only when needed (one, use it, discard it, next...). Even `range(1, 10000000000)` would use the same tiny memory! 🎯

> ✅ This is why iterators are the **efficient** way to loop through data.

## 3️⃣ Iterable vs Iterator vs Iteration

Three words that sound similar but mean different things:

| Term | Meaning |
|------|---------|
| **Iteration** | The *process* of going through items one by one |
| **Iterable** | An *object you can loop over* (list, tuple, set, dict, string, range) |
| **Iterator** | The *object that performs* the iteration (fetches items one by one) |

### How are iterable and iterator connected?
An **iterable gives you an iterator** when you call `iter()` on it.

In [ ]:
l = [1, 2, 3]
print(type(l))          # a list is an ITERABLE
print(type(iter(l)))    # iter() extracts its ITERATOR

<class 'list'>
<class 'list_iterator'>


### 📌 Two key facts
- **Every iterator is also an iterable** (you can loop over an iterator too).
- **Not every iterable is an iterator.** A list is iterable, but it is **not** an iterator — because a list stores all its items in memory at once, while a true iterator does not.

## 4️⃣ How to Check: Iterable or Iterator?

Every object has hidden **magic methods** (you can list them with `dir()`). The trick:
- An **iterable** has an **`__iter__`** method.
- An **iterator** has **both** **`__iter__`** and **`__next__`**.

### Test 1: is it iterable? (does it have `__iter__`?)

An integer is **not** iterable — looping over it fails.

In [ ]:
a = 2
for i in a:      # ❌ TypeError: 'int' object is not iterable
    print(i)

In [ ]:
# check by listing its methods — there is NO __iter__ for an int
a = 2
print("__iter__" in dir(a))   # False -> not iterable

False


A **tuple**, **set**, and **dict** all have `__iter__`, so they *are* iterable.

In [ ]:
print("__iter__" in dir((1, 2, 3)))    # tuple -> True
print("__iter__" in dir({1, 2, 3}))    # set   -> True
print("__iter__" in dir({1: 2, 3: 4})) # dict  -> True

True
True
True


### Test 2: is it an iterator? (does it have `__next__`?)

A **list is iterable but NOT an iterator** — it has `__iter__` but no `__next__`.

In [ ]:
L = [1, 2, 3]
print("__iter__" in dir(L))    # True  -> iterable
print("__next__" in dir(L))    # False -> but NOT an iterator

True
False


But when we extract its **iterator** with `iter()`, that object has **both** `__iter__` and `__next__`.

In [ ]:
L = [1, 2, 3]
iter_L = iter(L)               # get the iterator
print("__iter__" in dir(iter_L))   # True
print("__next__" in dir(iter_L))   # True -> it IS an iterator

True
True


## 5️⃣ How a `for` Loop *Really* Works

A `for` loop has no counter and no condition — yet it perfectly walks through any iterable. How? Behind the scenes it does **two things**:

1. **Get the iterator** from the iterable, using `iter()`.
2. **Repeatedly call `next()`** on that iterator to get each item — until `next()` raises a **`StopIteration`** error, which signals "no more items."

Let's do those steps manually.

In [ ]:
num = [1, 2, 3]

iter_num = iter(num)   # Step 1: get the iterator

print(next(iter_num))  # Step 2: get the next item -> 1
print(next(iter_num))  # -> 2
print(next(iter_num))  # -> 3
# print(next(iter_num))  # a 4th call would raise StopIteration (no items left)

1
2
3


So a `for` loop is really just: **get the iterator → keep calling `next()` → stop on `StopIteration`.**

## 6️⃣ Building Our Own `for` Loop

Now that we know the secret, we can build our **own** version of the `for` loop! We get the iterator, then keep calling `next()` inside a `while True` loop, breaking when `StopIteration` appears.

In [ ]:
def my_for_loop(iterable):
    iterator = iter(iterable)     # Step 1: get the iterator
    while True:
        try:
            print(next(iterator)) # Step 2: get the next item
        except StopIteration:     # no items left -> stop
            break

And it works on **every** kind of iterable — list, range, tuple, set, and dict!

In [ ]:
a = [1, 2, 3]
b = range(1, 6)
c = (1, 2, 3)
d = {1, 2, 3}
e = {0: 1, 1: 1}

my_for_loop(a)
my_for_loop(b)
my_for_loop(c)
my_for_loop(d)
my_for_loop(e)   # looping a dict gives its keys

1
2
3
1
2
3
4
5
1
2
3
1
2
3
0
1


## 🔍 A Confusing Point — `iter()` on an iterator returns itself

Calling `iter()` on an **iterable** gives a **new iterator**. But calling `iter()` on an **iterator** gives back **the same iterator** (not a new one). We can prove it with `id()` — the memory addresses are identical.

In [ ]:
num = [1, 2, 3]
iter_obj = iter(num)
print(id(iter_obj), "-> address of iterator 1")

iter_obj2 = iter(iter_obj)   # iter() on an iterator
print(id(iter_obj2), "-> address of iterator 2 (SAME!)")

135629524492080 -> address of iterator 1
135629524492080 -> address of iterator 2 (SAME!)


## 7️⃣ Building Our Own `range()` ⭐

Why learn all this? So you can **add iteration to your own objects** — make your own class that you can loop over!

To make a class loopable, it needs an **`__iter__`** method (making it *iterable*) that returns an **iterator object**. The iterator class needs **`__iter__`** and **`__next__`** (which raises `StopIteration` when done).

We build two classes: `mera_range` (the iterable) and `mera_range_iterator` (the iterator).

> 🛠️ **Fixed from the original:** the iterator class name had a typo (`mera_rangge_iterator`, with a double 'g') — corrected to `mera_range_iterator`.

In [ ]:
# 1) the ITERABLE class
class mera_range:
    def __init__(self, start, end):
        self.start = start
        self.end = end

    def __iter__(self):                      # makes it iterable
        return mera_range_iterator(self)     # returns an iterator object

In [ ]:
# 2) the ITERATOR class
class mera_range_iterator:
    def __init__(self, iterable_obj):
        self.iterable = iterable_obj

    def __iter__(self):        # an iterator returns itself
        return self

    def __next__(self):        # produces the next value, or stops
        if self.iterable.start >= self.iterable.end:
            raise StopIteration
        current = self.iterable.start
        self.iterable.start += 1
        return current

Now our custom object works with a normal `for` loop, just like the real `range()`!

In [ ]:
for i in mera_range(1, 11):
    print(i)

1
2
3
4
5
6
7
8
9
10


In [ ]:
x = mera_range(1, 11)
print(type(x))        # our own iterable class
print(type(iter(x)))  # our own iterator class

<class '__main__.mera_range'>
<class '__main__.mera_range_iterator'>


## 🌍 Where Is This Used in Real Life?

Iterators are the key to handling **data too big to fit in memory**.

🧠 **Real example (Data Science / ML):** imagine you have **20 GB of images** to process, but your computer only has **8 GB of RAM**. You *can't* load all the images at once. With an **iterator**, you load **one image at a time**, convert it, use it, discard it, then load the next — so you only ever hold one image in memory. This lets you process **any** amount of data, no matter how big.

> 💡 That's the big benefit of iterators: **at any moment, only ONE item is in memory.**

But writing a whole iterator class (like our `mera_range`) is a lot of work. There's a **much easier way** to create iterators — **Generators**, which we'll learn next. 👇

## 8️⃣ Generators — The Easy Way to Make Iterators ⭐

Did you notice that building `mera_range` needed **two whole classes** and several magic methods? That's a lot of code just to loop over something.

A **generator** does the same job in a **simple function** — no classes, no `__iter__`, no `__next__`, no `StopIteration`. A generator is just a **special function that produces items one at a time** using a new keyword: **`yield`**.

### The `yield` keyword

`yield` is like `return`, but with a superpower: instead of ending the function, it **pauses** it and **remembers** where it stopped. The next time you ask for a value, it **resumes** right after the `yield`.

In [ ]:
# A simple generator function
def gen():
    yield 1     # pause here and hand back 1
    yield 2     # next time, resume here and hand back 2
    yield 3

g = gen()          # calling it gives us a generator (an iterator!)
print(next(g))     # 1
print(next(g))     # 2
print(next(g))     # 3
# print(next(g))   # a 4th call raises StopIteration

1
2
3


### The same `mera_range`, but as a generator

Compare this to the two-class version earlier. **This does exactly the same thing in 4 lines!** 🤯

In [ ]:
# Our own range() — the generator way (no classes needed!)
def mera_range(start, end):
    while start < end:
        yield start        # hand back the current number, then pause
        start += 1

for i in mera_range(1, 11):
    print(i)

1
2
3
4
5
6
7
8
9
10


> 💡 A generator function automatically has `__iter__` and `__next__` behind the scenes — so it **is** an iterator. Python builds all the machinery for you.

### Generator Expressions

Remember list comprehension `[i**2 for i in range(5)]`? If you swap the **square brackets `[]` for round brackets `()`**, you get a **generator expression** — a one-line generator that produces values lazily (on demand).

In [ ]:
# a generator expression — note the round brackets ( )
gen = (i ** 2 for i in range(1, 6))
for x in gen:
    print(x)

1
4
9
16
25


### Generators save memory too

Just like `range`, a generator holds only **one item at a time**. Compare a list comprehension (stores everything) vs a generator expression (stores nothing until asked):

In [ ]:
import sys
list_comp = [i for i in range(100000)]     # stores all 100,000 numbers
gen_exp   = (i for i in range(100000))     # stores nothing — generates on demand

print(round(sys.getsizeof(list_comp) / 1024, 2), "KB  (list comprehension)")
print(round(sys.getsizeof(gen_exp) / 1024, 3), "KB  (generator expression)")

782.21 KB  (list comprehension)
0.109 KB  (generator expression)


😲 Again — **782 KB vs 0.1 KB!** Generators give you all the memory savings of an iterator, with the simplicity of a normal function.

### Iterator vs Generator — which to use?

| | Iterator (class) | Generator (function) |
|--|------------------|----------------------|
| How you write it | a class with `__iter__` + `__next__` | a function with `yield` |
| Amount of code | a lot | very little |
| Memory | efficient (one item at a time) | efficient (one item at a time) |
| When to use | when you need full control | almost always — it's the easy default |

> ✅ **Bottom line:** generators are the **easiest and most common** way to create iterators. Reach for them whenever you need to produce a sequence lazily.

---

## 🎯 Summary — Quick Revision

You now understand what powers every loop in Python! 🎉

| Concept | Key idea |
|---------|----------|
| **Iteration** | Going through items one by one |
| **Iterable** | An object you can loop over (has `__iter__`) |
| **Iterator** | The object doing the work (has `__iter__` **and** `__next__`) |
| **Memory saver** | An iterator holds only ONE item at a time (`range` vs a list) |
| **`iter()` / `next()`** | Get an iterator / get its next item |
| **`StopIteration`** | The signal that there are no more items |
| **How `for` works** | `iter()` once, then `next()` until `StopIteration` |
| **Custom iteration** | Add `__iter__` + `__next__` to make your own class loopable |
| **Generator** | A function with `yield` — the *easy* way to make an iterator |
| **`yield`** | Like `return`, but pauses and resumes the function |
| **Generator expression** | `(i for i in ...)` — round brackets instead of `[ ]` |

---

### 📝 Try It Yourself (Practice Questions)

1. Use `iter()` and `next()` to manually pull items out of a string.
2. Check whether a `float`, a `str`, and a `range` object are iterable (look for `__iter__`).
3. Build your own for-loop function that also **counts** how many items it printed.
4. Write a **generator** `countdown(n)` that yields `n, n-1, ... 1`.
5. Write a **generator expression** that produces the cubes of numbers 1 to 10.

> 💪 **Remember:** iterators feel abstract, but they're the reason Python can process huge files and datasets on a normal laptop. This is a real superpower for Data Science!

---

**🎓 That completes Week 4 — Advanced Python! (File Handling, Exceptions, Decorators & Iterators.)**